### PR TITLE
resolve ninja warnings

### PR DESCRIPTION
**Pull request by Lafydev**
Resolve ninja warnings (about deprecated widgets...)

### CODE REVIEW

**Comment by Lafydev** [comment]
warnings namarupa

**Comment by Lafydev** [comment]
solve issue #4

**Comment by Lafydev** [comment]
screenshot with submenus
namarupa

**Comment by Lafydev** [comment]
Please can you tell me what you are waiting from the settings button ?

**Comment by lainsce** [comment]
The original idea was a Switchboard plug where one could hide or rearrange the icons inside the indicator popover.

**Comment by Lafydev** [comment]
> The original idea was a Switchboard plug where one could hide or rearrange the icons inside the indicator popover.

Hum, very ambitious but I'll stop here, perhaps we just need a new .deb file for easy install.

File: src/AyatanaIndicator.vala [hunk 1]
```diff
@@ -34,9 +35,13 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
     private Gtk.Popover popover;
 
     private Gee.HashMap<Gtk.Widget, Gtk.Widget> menu_map;
+	private Gee.HashMap<Gtk.Widget, Gtk.Widget> submenu_map;
 
-    const int MAX_ICON_SIZE = 22;
+	 const int MAX_ICON_SIZE = 22;
 
+	//group radiobuttons
+    private Gtk.RadioButton? group_radio=null ;
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]: Time for some code styling changes to fit elementary guidelines.

```suggestion
    private Gtk.RadioButton? group_radio = null;
```

File: src/AyatanaIndicator.vala [hunk 2]
```diff
@@ -145,8 +141,7 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
                      * the menu is popuped
                      */
                     reloaded = true;
-                    entry.menu.popup (null, null, null, 0, Gtk.get_current_event_time ());
-                    entry.menu.popdown ();
+                    entry.menu.popup_at_widget (icon.parent,  0, 0);        //entry.menu.popdown ();
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]:
```suggestion
                    entry.menu.popup_at_widget (icon.parent, 0, 0);
```

File: src/AyatanaIndicator.vala [hunk 3]
```diff
@@ -220,14 +215,14 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
             button.no_show_all = true;
             button.hide ();
         });
-        item.state_changed.connect ((type) => {
-            button.set_state (item.get_state ());
+        item.state_flags_changed.connect ((type) => {
+            button.set_state_flags (item.get_state_flags (),true);
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]:
```suggestion
            button.set_state_flags (item.get_state_flags (), true);
```

File: src/AyatanaIndicator.vala [hunk 4]
```diff
@@ -244,14 +239,20 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
          * get item type from atk accessibility
          * 34 = MENU_ITEM  8 = CHECKBOX  32 = SUBMENU 44 = RADIO
          */
+		const int ATK_CHECKBOX =8;
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]:
```suggestion
		const int ATK_CHECKBOX = 8;
```

File: src/AyatanaIndicator.vala [hunk 5]
```diff
@@ -244,14 +239,20 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
...
+		const int ATK_RADIO =44;
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]:
```suggestion
		const int ATK_RADIO = 44;
```

File: src/AyatanaIndicator.vala [hunk 6]
```diff
@@ -244,14 +239,20 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
...
         Value val = Value (typeof (int));
         atk.get_property ("accessible_role", ref val);
         var item_type = val.get_int ();
 
-        var state = item.get_state ();
+        var state = item.get_state_flags ();
         var active = (item as Gtk.CheckMenuItem).get_active ();
 
+		//RAZ group_radio
+       group_radio = ( item_type == ATK_RADIO)? group_radio:null;
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]:
```suggestion
       group_radio = (item_type == ATK_RADIO)? group_radio:null;
```

File: src/AyatanaIndicator.vala [hunk 7]
```diff
@@ -264,27 +265,50 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
...
+        if (item_type == ATK_CHECKBOX) {
             var button = new Wingpanel.Widgets.Switch (label, active);
             button.get_switch ().state_set.connect ((b) => {
                 (item as Gtk.CheckMenuItem).set_active (b);
                 //  close ();
 
                 return false;
             });
-            button.set_state (state);
+            button.set_state_flags (state,false);
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]:
```suggestion
            button.set_state_flags (state, false);
```

File: src/AyatanaIndicator.vala [hunk 8]
```diff
@@ -264,27 +265,50 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
...
 
             connect_signals (item, button);
             (item as Gtk.CheckMenuItem).toggled.connect (() => {
-                button.set_active ((item as Gtk.CheckMenuItem).get_active ());
+                button.active =((item as Gtk.CheckMenuItem).get_active ());
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]:
```suggestion
                button.active = ((item as Gtk.CheckMenuItem).get_active ());
```

File: src/AyatanaIndicator.vala [hunk 9]
```diff
@@ -264,27 +265,50 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
...
+                    item.activate ();
+                });
+			
+			button.activate.connect (() => {
+				item.activate();
+               (item as Gtk.RadioMenuItem).set_active (button.get_active());
+            });
+			
+			return button;
+		}
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]:
```suggestion
		if (item_type == ATK_RADIO) {
		    var button = new Gtk.RadioButton.with_label_from_widget (group_radio,label) {
		       margin = 5,
		       margin_start = 10
		    };
		    if (group_radio == null) {
		        group_radio = button;
		    }
		    
	            button.set_active(active);
			
		    // do not remove
                  button.clicked.connect (() => {
                        item.activate ();
                  });
			
		    button.activate.connect (() => {
			item.activate();
                       (item as Gtk.RadioMenuItem).set_active (button.get_active ());
                   });
			
		    return button;
		}
```

File: src/AyatanaIndicator.vala [hunk 10]
```diff
@@ -294,53 +318,73 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
                     warning (e.message);
                 }
             }
-
+			button = new Gtk.ModelButton();
+			button.text=label;
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]:
```suggestion
			button.text = label;
```

File: src/AyatanaIndicator.vala [hunk 11]
```diff
@@ -294,53 +318,73 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
...
             if (image != null && image.pixbuf != null) {
-                button = new Wingpanel.Widgets.Button (label);
-                (button as Wingpanel.Widgets.Button).set_pixbuf (image.pixbuf);
-            } else {
-                button = new Wingpanel.Widgets.Button (label);
-            }
+                (button as Gtk.ModelButton).icon= (image.pixbuf);
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]:
```suggestion
                (button as Gtk.ModelButton).icon = (image.pixbuf);
```

File: src/AyatanaIndicator.vala [hunk 12]
```diff
@@ -294,53 +318,73 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
...
+            } 
             (item as Gtk.CheckMenuItem).notify["label"].connect (() => {
-                (button as Wingpanel.Widgets.Button).set_caption ((item as Gtk.MenuItem).get_label ().replace ("_", ""));
+                (button as Gtk.ModelButton).text = ((item as Gtk.MenuItem).get_label ().replace ("_", ""));
             });
 
-            button.set_state (state);
+            button.set_state_flags (state,true);
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]:
```suggestion
            button.set_state_flags (state, true);
```

File: src/AyatanaIndicator.vala [hunk 13]
```diff
@@ -294,53 +318,73 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
...
                 scroll_sub.set_policy (Gtk.PolicyType.AUTOMATIC, Gtk.PolicyType.AUTOMATIC);
-                var sub_stack = new Gtk.Grid ();
-                scroll_sub.add (sub_stack);
-                var back_button = new Wingpanel.Widgets.Button (_("Back"));
+                scroll_sub.add (sub_list);
+				
+                var back_button = new Gtk.ModelButton();
+				back_button.text=(_("Back"));
+				back_button.inverted=true;
+				back_button.menu_name="main_grid";
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]:
```suggestion
		    back_button.text = (_("Back"));
		    back_button.inverted = true;
		    back_button.menu_name = "main_grid";
```

File: src/AyatanaIndicator.vala [hunk 14]
```diff
@@ -294,53 +318,73 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
...
-                submenu.popdown ();
+				
+				submenu.remove.connect ((item)=> {
+					var w = menu_map.get (item);
+
+        			if (w != null) {
+            			sub_list.remove (w);
+            			submenu_map.unset (item);
+        			}	
+				});
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]:
```suggestion
		     submenu.remove.connect ((item) => {
			 var w = menu_map.get (item);

        		 if (w != null) {
            		     sub_list.remove (w);
            		     submenu_map.unset (item);
        		 }	
		    });
```

File: src/AyatanaIndicator.vala [hunk 15]
```diff
@@ -294,53 +318,73 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
...
+               
                 main_stack.add (scroll_sub);
-                button = new SubMenuButton (label);
+				//Button opening the submenu
+                button = new Gtk.ModelButton();
+				button.text=label;
+				button.menu_name="submenu";
```
Comment (lainsce) [inline src/AyatanaIndicator.vala]:
```suggestion
		      button.text = label;
		      button.menu_name = "submenu";
```

File: src/Indicator.vala [hunk 16]
```diff
@@ -22,7 +22,8 @@ public class AyatanaCompatibility.MetaIndicator : Wingpanel.Indicator {
     private Gtk.Label label;
     private Gtk.Stack stack;
     private Gtk.Box? main_box = null;
-
+	public int cpt=0; /* count indicators */
```
Comment (lainsce) [inline src/Indicator.vala]:
```suggestion
	public int cpt = 0; /* count indicators */
```

File: src/Indicator.vala [hunk 17]
```diff
@@ -58,66 +59,76 @@ public class AyatanaCompatibility.MetaIndicator : Wingpanel.Indicator {
         if (blacklist.contains (indicator.name_hint ())) {
             return;
         }
-
-        get_widget ();
+		 if (main_box == null) {init_main_box();}
```
Comment (lainsce) [inline src/Indicator.vala]:
```suggestion
		    if (main_box == null) {
		        init_main_box ();
		    }
```

File: src/Indicator.vala [hunk 18]
```diff
@@ -58,66 +59,76 @@ public class AyatanaCompatibility.MetaIndicator : Wingpanel.Indicator {
...
-        switch_stack (false);
+        //get_widget ();
 
         foreach (var fbc in box.get_children ()) {
             var child = (Gtk.Widget)fbc;
+			//see what append when two indicators have the same name ? 
             if (child is Indicator && ((Indicator)child).code_name == indicator.code_name) {
                 child.destroy ();
+                cpt--;
+				if (cpt==0) {switch_stack (false);}
```
Comment (lainsce) [inline src/Indicator.vala]:
```suggestion
				if (cpt == 0) {
				    switch_stack (false);
				}  
```

File: src/Indicator.vala [hunk 19]
```diff
@@ -58,66 +59,76 @@ public class AyatanaCompatibility.MetaIndicator : Wingpanel.Indicator {
...
             main_box.add (new Wingpanel.Widgets.Separator ());
             main_box.add (settings_btn);
-        }
+        
 
         return main_box;
+	}
+
+    public override Gtk.Widget? get_widget () {
+		if (main_box == null) {init_main_box();}
```
Comment (lainsce) [inline src/Indicator.vala]:
```suggestion
		if (main_box == null) {
		    init_main_box ();
		}
```

File: src/Indicator.vala [hunk 20]
```diff
@@ -137,6 +148,7 @@ public class AyatanaCompatibility.MetaIndicator : Wingpanel.Indicator {
         foreach (var entry in get_restrictions_from_file (blacklist_file)) {
             blacklist.add (entry);
         }
+		blacklist.add("nm-applet"); //old network indicator (duplicate)
```
Comment (lainsce) [inline src/Indicator.vala]:
```suggestion
		blacklist.add ("nm-applet"); //old network indicator (duplicate)
```

File: src/Indicator.vala [hunk 21]
```diff
@@ -160,16 +172,23 @@ public class AyatanaCompatibility.MetaIndicator : Wingpanel.Indicator {
...
+    // TODO: Plug for Namarupa.  
+    private void show_settings () {
+		/* temporary used for informations */
+		string msg= cpt.to_string() + " item(s) \n";
+		var msgdial = new Gtk.MessageDialog(null,Gtk.DialogFlags.MODAL,
+        Gtk.MessageType.INFO, Gtk.ButtonsType.CLOSE,
+        msg);
+		msgdial.set_title("Information");
+		msgdial.run();
+		msgdial.destroy(); //dialog without parent =>not perfectly closed
```
Comment (lainsce) [inline src/Indicator.vala]:
```suggestion
		string msg = cpt.to_string () + " item(s)\n";
		var msgdial = new Gtk.MessageDialog (null,
		                                     Gtk.DialogFlags.MODAL,
                                                   Gtk.MessageType.INFO,
                                                   Gtk.ButtonsType.CLOSE,
                                                   msg);
		msgdial.set_title (_("Information"));
		msgdial.run ();
		msgdial.destroy (); // dialog without parent => not perfectly closed
```

**Review by lainsce — CHANGES_REQUESTED** [review]
Left some parts to fix to fit code style guidelines.

**Comment by lainsce** [comment]
OK all seems good. Merging.

File: src/AyatanaIndicator.vala [hunk 22]
```diff
@@ -285,25 +285,27 @@ public class AyatanaCompatibility.Indicator : IndicatorButton {
 
 		//RADIO BUTTON
 		if (item_type == ATK_RADIO) {
-					
-			var button= new Gtk.RadioButton.with_label_from_widget(group_radio,label);
-			if (group_radio==null) {group_radio=button;}
-			button.margin = 5;
-			button.set_margin_start(10);
+		    var button = new Gtk.RadioButton.with_label_from_widget (group_radio,label) {
```
Comment (Lafydev) [inline src/AyatanaIndicator.vala:288]: I think it's not a } but ;  l.288  ? I don't understand this change

**Comment by Lafydev** [comment]
Thanks a lot for merging the changes. 
Please can you change line 288 it is a semi-column at the end of the line.

**Pull request merged**